optuna/_hypervolume/box_decomposition.py

```diff
@@ -0,0 +1,144 @@
+        # NOTE(nabenabe): The indices of `u` with `True` in update. Each `u` may yield `True`
+        # multiple times for different indices `j`.
+        ubs_indices_to_update = np.tile(np.arange(n_bounds)[:, np.newaxis], n_objectives)[update]
+        # The dimension `j` for each `u` s.t. `\hat{z}_j \geq \max_{k \neq j}{z_j^k(u)}`.
+        dimensions_to_update = np.tile(objective_indices, (n_bounds, 1))[update]
+        assert ubs_indices_to_update.size == dimensions_to_update.size
+        indices_for_sweeping = np.arange(dimensions_to_update.size)
+        # The last Eq in Page 5.
+        new_dps = dominated_dps[ubs_indices_to_update]
+        new_dps[indices_for_sweeping, dimensions_to_update] = sol
+        # Line 3 of Alg. 2. `sol[dimensions_to_update]` is equivalent to `\bar{z}_j`.
+        new_ubs = ubs[is_dominated][ubs_indices_to_update]
+        new_ubs[indices_for_sweeping, dimensions_to_update] = sol[dimensions_to_update]
+        return np.vstack([ubs[~is_dominated], new_ubs]), np.vstack([dps[~is_dominated], new_dps])
+
+    upper_bound_set = np.asarray([ref_point])  # Line 1 of Alg. 2.
+    def_points = np.full((1, n_objectives, n_objectives), -np.inf)  # z^k(z^r) = \hat{z}^k
+    def_points[0, objective_indices, objective_indices] = ref_point  # \hat{z}^k is a dummy point.
+    for solution in sorted_pareto_sols:  # NOTE(nabenabe): Sorted must be fulfilled.
+        upper_bound_set, def_points = update(solution, upper_bound_set, def_points)
+
+    return upper_bound_set, def_points
+
+
+def _get_box_bounds(
+    def_points: np.ndarray, upper_bound_set: np.ndarray, ref_point: np.ndarray
+) -> np.ndarray:
+    # Eq. (2) of Lacour17.
+    n_objectives = upper_bound_set.shape[-1]
+    bounds = np.empty((2, *upper_bound_set.shape))
+    bounds[0, :, 0] = def_points[:, 0, 0]
+    bounds[1, :, 0] = ref_point[0]
+    row, col = np.diag_indices(n_objectives - 1)
+    bounds[0, :, 1:] = np.maximum.accumulate(def_points, axis=-2)[:, row, col + 1]
+    bounds[1, :, 1:] = upper_bound_set[:, 1:]
+    not_empty = ~np.any(bounds[1] <= bounds[0], axis=-1)  # Remove [inf, inf] or [-inf, -inf].
+    return bounds[:, not_empty]
+
+
+def _get_non_dominated_box_bounds(
+    sorted_pareto_sols: np.ndarray, ref_point: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:  # (n_bounds, n_objectives) and (n_bounds, n_objectives)
+    # The calculation of u[k] and l[k] in the paper: https://arxiv.org/abs/2006.05078
+    # NOTE(nabenabe): The paper handles maximization problems, but we consider minimization here.
+    _upper_bound_set = _get_upper_bound_set(sorted_pareto_sols, ref_point)[0]
+    upper_bound_set = _upper_bound_set[np.argsort(_upper_bound_set[:, 0])[::-1]]
+    # Flip the sign and use upper_bound_set as the Pareto solutions. Then we can calculate the
+    # lower bound set as well.
+    point_at_infinity = np.full_like(ref_point, np.inf)
+    neg_upper_bound_set, neg_def_points = _get_upper_bound_set(-upper_bound_set, point_at_infinity)
```

Comment:
It’s true that passing `-upper_bound_set` after sorting `_upper_bound_set` in descending order results in an ascending order due to a coincidental symmetry. However, this behavior seems to rely implicitly on that symmetry, which could easily break if the sorting logic is refactored in the future. To make the implementation more robust, it might be better to explicitly enforce ascending order using `np.lexsort` after negation. What do you think?

Reply:
I am not sure if I addressed your comment, but I made some modifications.

Reply:
I will check it!